Refs #18597 -- Refactor new hooks for BaseModelFormSet and BaseInlineFormSet

shjohnson-pi opened the pull request:
Added two new hooks:

- BaseModelFormSet.get_instance_by_queryset_index
- BaseInlineFormSet.prepare_queryset

Added documentation showing how new hooks can be used to allow an inline formset to use prefetch_related. Added a test for this scenario.

---
Context:

* https://code.djangoproject.com/ticket/5372 Inline forms should use prefetch_related to reduce queries
* https://code.djangoproject.com/ticket/18597 prefetch_related doesn't work as expected for inline forms
* https://github.com/django/django/pull/14188 [agalazis](https://github.com/agalazis) started it, but it needed docs/tests. Last worked on November 2021
    * My pull request is based on his, but with the added docs and tests

Because it's been ~2 years since there's been progress on these tickets, I'm sure extra time will be needed to determine if this is still the best direction to fix the inline/cache problem.

Also, this is my first major pull request for Django, so all feedback is welcome. In particular my documentation and tests may need changes.

felixxm commented on docs/topics/forms/modelforms.txt:
```diff
@@ -1311,6 +1311,51 @@ Then when you create your inline formset, pass in the optional argument
     >>> author = Author.objects.get(name="Mike Royko")
     >>> formset = BookFormSet(instance=author)
 
+.. _inline-formset-interaction-with-prefetching:
```
`get_instance_by_queryset_index()` should also be documented under `models.BaseModelFormSet` [docs](https://docs.djangoproject.com/en/5.0/topics/forms/modelforms/#model-formsets). Is it useful in itself?

shjohnson-pi replied:
I'm not sure how to mention `get_instance_by_queryset_index()` in the 'Model formsets' section.
I can't think of a reason to override that method on it's own.

felixxm commented on docs/topics/forms/modelforms.txt:
```diff
@@ -1311,6 +1311,51 @@ Then when you create your inline formset, pass in the optional argument
+
+.. versionadded:: 5.1
+
+    The ``get_instance_by_queryset_index`` and ``prepare_queryset`` hooks were added.
+
```
This entire section is new so this comment is unnecessary
```suggestion
```

felixxm commented on docs/topics/forms/modelforms.txt:
```diff
@@ -1311,6 +1311,51 @@ Then when you create your inline formset, pass in the optional argument
+
+    The ``get_instance_by_queryset_index`` and ``prepare_queryset`` hooks were added.
+
+You'll need to create a subclass to take advantage of
```
Reading this section, it's not clear for me what we want to achieve. We should start from describing what we're trying to do here.

felixxm commented on docs/topics/forms/modelforms.txt:
```diff
@@ -1311,6 +1311,51 @@ Then when you create your inline formset, pass in the optional argument
+                    return queryset
+            return super().prepare_queryset(queryset)
+
+Note that you'll also need to ensure that the related object has a default sort::
```
My immediate question is "why?".

felixxm commented on tests/model_formsets/tests.py:
```diff
@@ -1142,6 +1144,55 @@ def test_inline_formsets_with_wrong_fk_name(self):
             inlineformset_factory(Author, Book, fields="__all__", fk_name="title")
 
+    def test_inline_formsets_with_prefetch_related(self):
+        """
+        Override BaseInlineFormSet.get_instance_by_queryset_index and
+        BaseInlineFormSet.prepare_queryset (#18597).
+        """
```
This docstring doesn't describe anything, I'd remove it.

felixxm commented on tests/model_formsets/tests.py:
```diff
@@ -1142,6 +1144,55 @@ def test_inline_formsets_with_wrong_fk_name(self):
+        BaseInlineFormSet.prepare_queryset (#18597).
+        """
+        author = Author.objects.create(pk=1, name="Charles Baudelaire")
+        Book.objects.create(pk=1, author=author, title="Les Paradis Artificiels")
+        Book.objects.create(pk=2, author=author, title="Les Fleurs du Mal")
+        Book.objects.create(pk=3, author=author, title="Flowers of Evil")
```
Please don't hardcode PKs.
```suggestion
        Book.objects.create(author=author, title="Les Paradis Artificiels")
        Book.objects.create(author=author, title="Les Fleurs du Mal")
        Book.objects.create(author=author, title="Flowers of Evil")
```

felixxm commented on tests/model_formsets/tests.py:
```diff
@@ -1142,6 +1144,55 @@ def test_inline_formsets_with_wrong_fk_name(self):
+        # Check the default issues one query
+        BookFormSet = inlineformset_factory(Author, Book, fields=["title"])
+
+        with CaptureQueriesContext(connection) as ctx:
+            formset = BookFormSet(instance=author)
+            self.assertEqual(len(formset.forms), 6)
+        self.assertEqual(len(ctx), 1)  # Only 1 query
```
You can use `self.assertNumQueries()`:
```suggestion
        with self.assertNumQueries(1):
            formset = BookFormSet(instance=author)
            self.assertEqual(len(formset.forms), 6
```

felixxm commented on tests/model_formsets/tests.py:
```diff
@@ -1142,6 +1144,55 @@ def test_inline_formsets_with_wrong_fk_name(self):
+        Book.objects.create(pk=2, author=author, title="Les Fleurs du Mal")
+        Book.objects.create(pk=3, author=author, title="Flowers of Evil")
+
+        # Check the default issues one query
```
To add more explanation, comments should state the expected behavior and omit prefixes like "Check" since all tests check things. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

felixxm commented on tests/model_formsets/tests.py:
```diff
@@ -1142,6 +1144,55 @@ def test_inline_formsets_with_wrong_fk_name(self):
+        author_with_books = Author.objects.prefetch_related("book_set").get(
+            name="Charles Baudelaire"
+        )
+        with CaptureQueriesContext(connection) as ctx:
+            formset = BookFormSet(instance=author_with_books)
+            self.assertEqual(len(formset.forms), 6)
+        self.assertFalse(ctx.captured_queries)  # No queries, prefetch_related is used
```
```suggestion
        with assertNumQueries(0):
            formset = BookFormSet(instance=author_with_books)
            self.assertEqual(len(formset.forms), 6)
```

felixxm commented on tests/model_formsets/tests.py:
```diff
@@ -1142,6 +1144,55 @@ def test_inline_formsets_with_wrong_fk_name(self):
+        self.assertFalse(ctx.captured_queries)  # No queries, prefetch_related is used
+
+        # Again but without prefetch_related
+        with CaptureQueriesContext(connection) as ctx:
+            formset = BookFormSet(instance=author)
+            self.assertEqual(len(formset.forms), 6)
+        self.assertEqual(len(ctx), 1)  # Only 1 query
```
```suggestion
        with assertNumQueries(1):
            formset = BookFormSet(instance=author)
            self.assertEqual(len(formset.forms), 6)
```

felixxm commented on docs/releases/5.1.txt:
```diff
@@ -177,6 +177,12 @@ Forms
+* New hooks were added to :class:`~django.forms.models.BaseModelFormSet` and
+  :class:`~django.forms.models.BaseInlineFormSet` to allow custom subclasses to
+  utilize :meth:`~django.db.models.query.QuerySet.prefetch_related`
+  (:ticket:`18597`). See :ref:`inline-formset-interaction-with-prefetching` for
```
Ticket number is unnecessary here
```suggestion
  See :ref:`inline-formset-interaction-with-prefetching` for
```

felixxm reviewed (commented):
@shjohnson-pi Thanks :+1: I left initial comments.

shjohnson-pi committed:
Refs #18597 -- Refactor new hooks for BaseModelFormSet and BaseInlineFormSet.

Added two new hooks:

- BaseModelFormSet.get_instance_by_queryset_index
- BaseInlineFormSet.prepare_queryset

Added documentation showing how new hooks can be used to allow an inline
formset to use prefetch_related. Added a test for this scenario.

shjohnson-pi commented:
Thank you for the detailed feedback felixxm.

I've made the following changes and rebased.

- Removed redundant sentence.
- Explained goal of docs section.
- Explained why a sort is required.
- Removed redundant test docstring.
- Removed hardcoded pks in test.
- Used `self.assertNumQueries()`. (Looks much better!)
- Tweaked test comments.
- Removed link to ticket in release notes.
